Named arguments' behavior is not same to Scala 2

The following code fails in scala 2.12.2:
```scala
scala> def printName(first: String, last: String): Unit = {
     |   println(first + " " + last)
     | }
printName: (first: String, last: String)Unit

scala> printName("John", first = "Smith")
<console>:13: error: parameter 'first' is already specified at parameter position 1
       printName("John", first = "Smith")
                               ^
```

But compiles in dotty ([Scastie link](https://scastie.scala-lang.org/cK7KgMWPTMmwQ0bKq0eZ9w)):
```scala
object Main {
  def printName(first: String, last: String): Unit = {
    println(first + " " + last)
  }

  def main(args: Array[String]): Unit = {
    printName("John", first = "Smith")
  }
}

scastie: Connecting.
scastie: Connected.
sbt: [info] Compiling 1 Scala source to /tmp/scastie5500368941075143733/target/scala-0.1/classes...
sbt: [info] Running Main
Smith John
```

If the new behavior is intentional, the similar code (I use `last: String*` instead of `last: String`) fails in dotty ([Scastie link](https://scastie.scala-lang.org/lEpIbo05TCm7Tl8cQAoTtw)):
```scala
object Main {
  def printName(first: String, last: String*): Unit = {
    println(first + " " + last)
  }

  def main(args: Array[String]): Unit = {
    printName("John", first = "Smith")
  }
}

scastie: Connecting.
scastie: Connected.
sbt: [info] Compiling 1 Scala source to /tmp/scastie5500368941075143733/target/scala-0.1/classes...
sbt: [info] Running Main
sbt: [error] (run-main-10b) java.lang.NoClassDefFoundError: scala/$repeated$
sbt: java.lang.NoClassDefFoundError: scala/$repeated$
	at Main$.main(main.scala:7)
	at Main.main(main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
Caused by: java.lang.ClassNotFoundException: scala.$repeated$
	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at Main$.main(main.scala:7)
	at Main.main(main.scala)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
[trace] Stack trace suppressed: run 'last compile:run' for the full output.
sbt: java.lang.RuntimeException: Nonzero exit code: 1
	at scala.sys.package$.error(package.scala:27)
[trace] Stack trace suppressed: run 'last compile:run' for the full output.
sbt: [error] (compile:run) Nonzero exit code: 1
```

--- a/compiler/src/dotty/tools/dotc/typer/Applications.scala
+++ b/compiler/src/dotty/tools/dotc/typer/Applications.scala
@@ -260,19 +260,19 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
        *  1. `(args diff toDrop)` can be reordered to match `pnames`
        *  2. For every `(name -> arg)` in `nameToArg`, `arg` is an element of `args`
        */
-      def recur(pnames: List[Name], args: List[Trees.Tree[T]],
-                nameToArg: Map[Name, Trees.NamedArg[T]], toDrop: Set[Name]): List[Trees.Tree[T]] = pnames match {
+      def handleNamed(pnames: List[Name], args: List[Trees.Tree[T]],
+                      nameToArg: Map[Name, Trees.NamedArg[T]], toDrop: Set[Name]): List[Trees.Tree[T]] = pnames match {
         case pname :: pnames1 if nameToArg contains pname =>
           // there is a named argument for this parameter; pick it
-          nameToArg(pname) :: recur(pnames1, args, nameToArg - pname, toDrop + pname)
+          nameToArg(pname) :: handleNamed(pnames1, args, nameToArg - pname, toDrop + pname)
         case _ =>
           def pnamesRest = if (pnames.isEmpty) pnames else pnames.tail
           args match {
             case (arg @ NamedArg(aname, _)) :: args1 =>
               if (toDrop contains aname) // argument is already passed
-                recur(pnames, args1, nameToArg, toDrop - aname)
+                handleNamed(pnames, args1, nameToArg, toDrop - aname)
               else if ((nameToArg contains aname) && pnames.nonEmpty) // argument is missing, pass an empty tree
-                genericEmptyTree :: recur(pnames.tail, args, nameToArg, toDrop)
+                genericEmptyTree :: handleNamed(pnames.tail, args, nameToArg, toDrop)
               else { // name not (or no longer) available for named arg
                 def msg =
                   if (methodType.paramNames contains aname)
@@ -280,17 +280,26 @@ trait Applications extends Compatibility { self: Typer with Dynamic =>
                   else
                     s"$methString does not have a parameter $aname"
                 fail(msg, arg.asInstanceOf[Arg])
-                arg :: recur(pnamesRest, args1, nameToArg, toDrop)
+                arg :: handleNamed(pnamesRest, args1, nameToArg, toDrop)
               }
             case arg :: args1 =>
-              arg :: recur(pnamesRest, args1, nameToArg, toDrop) // unnamed argument; pick it
+              arg :: handleNamed(pnamesRest, args1, nameToArg, toDrop) // unnamed argument; pick it
             case Nil => // no more args, continue to pick up any preceding named args
               if (pnames.isEmpty) Nil
-              else recur(pnamesRest, args, nameToArg, toDrop)
+              else handleNamed(pnamesRest, args, nameToArg, toDrop)
           }
       }
-      val nameAssocs = for (arg @ NamedArg(name, _) <- args) yield (name, arg)
-      recur(methodType.paramNames, args, nameAssocs.toMap, Set())
+
+      def handlePositional(pnames: List[Name], args: List[Trees.Tree[T]]): List[Trees.Tree[T]] =
+        args match {
+          case (arg: NamedArg) :: _ =>
+            val nameAssocs = for (arg @ NamedArg(name, _) <- args) yield (name, arg)
+            handleNamed(pnames, args, nameAssocs.toMap, Set())
+          case arg :: args1 => arg :: handlePositional(pnames.tail, args1)
+          case Nil => Nil
+        }
+
+      handlePositional(methodType.paramNames, args)
     }
 
     /** Splice new method reference into existing application */

--- a/tests/neg/i2778.scala
+++ b/tests/neg/i2778.scala
@@ -1,0 +1,9 @@
+object Main {
+  def printName(first: String, last: String): Unit = {
+    println(first + " " + last)
+  }
+
+  def main(args: Array[String]): Unit = {
+    printName("John", first = "Smith") // error: parameter is already instantiated
+  }
+}

